makefiles/arch/mips.inc.mk

```diff
@@ -1,41 +1,68 @@
-ifndef MIPS_ELF_ROOT
-ifneq ($(BUILD_IN_DOCKER),1) #Don't error when BUILD_IN_DOCKER=1 as it _is_ set in DOCKER
-    $(error "Please set $$(MIPS_ELF_ROOT) and ensure $$(MIPS_ELF_ROOT)/bin is on your PATH")
-endif
-endif
-
 # Target triple for the build.
 export TARGET_ARCH ?= mips-mti-elf
 
 export ABI=32
 
-ifneq ($(BUILD_IN_DOCKER),1) #Don't error when BUILD_IN_DOCKER=1 as MIPS_ELF_ROOT _is_ set in DOCKER
-include $(MIPS_ELF_ROOT)/share/mips/rules/mipshal.mk
+# Portable 'lowercase' func.
+lc = $(subst A,a,$(subst B,b,$(subst C,c,$(subst D,d,$(subst E,e,$(subst F,f,$(subst G,g,$(subst H,h,$(subst I,i,$(subst J,j,$(subst K,k,$(subst L,l,$(subst M,m,$(subst N,n,$(subst O,o,$(subst P,p,$(subst Q,q,$(subst R,r,$(subst S,s,$(subst T,t,$(subst U,u,$(subst V,v,$(subst W,w,$(subst X,x,$(subst Y,y,$(subst Z,z,$1))))))))))))))))))))))))))
+
+# Default values for the linker script symbols listed below are
+# defined in the linker script.
+
+# These are linker script symbols that are prefixed with '__"
+priv_symbols = MEMORY_BASE MEMORY_SIZE STACK
+priv_symbols += ENABLE_XPA
+priv_symbols += FLUSH_TO_ZERO
+priv_symbols += FLASH_START APP_START FLASH_APP_START
+priv_symbols += ISR_VEC_SPACE ISR_VECTOR_COUNT
+
+comma := ,
+
+# A bit of makefile magic:
+# foreach symbol in overridable ld-symbols :
+#   If symbol has a value, produce a linker argument for that symbol.
+MIPS_HAL_LDFLAGS = $(foreach a,$(priv_symbols),$(if $($a),-Wl$(comma)--defsym$(comma)__$(call lc,$(a))=$($a)))
+
+ifeq ($(ROMABLE),1)
+MIPS_HAL_LDFLAGS += -T bootcode.ld
 endif
 
 # define build specific options
-export CFLAGS_CPU   = -EL -std=gnu99
+# Remove -std=gnu99 once the MIPS toolchain headers are updated to include upstream
+# newlib commit 81c17949f0419d1c4fee421c60987ea1149522ae
+# https://cygwin.com/git/gitweb.cgi?p=newlib-cygwin.git;a=commitdiff;h=81c17949f0419d1c4fee421c60987ea1149522ae
+# Otherwise we get an error about a missing declaration of strnlen in some parts.
+export CFLAGS += -std=gnu99
+export CFLAGS_CPU   = -EL -mabi=$(ABI)
+# Why not -fdata-sections?? /JN
 export CFLAGS_LINK  = -ffunction-sections -fno-builtin -fshort-enums #-fdata-sections
```

Comment:
I was having problems with the linker stripping out the pic32 configuration registers (which have to reside at a specific place in the final flash image), but this was actually due to a missing KEEP() attribute in the linker script, so this can be put back now.